components/layout/sequential.rs

```diff
@@ -75,27 +79,54 @@ pub fn build_display_list_for_subtree<'a>(flow_root: &mut Flow,
     let mut state = DisplayListBuildState::new(layout_context);
     flow_root.collect_stacking_contexts(&mut state);
 
-    let mut build_display_list = BuildDisplayList { state: state };
+    let mut build_display_list = BuildDisplayList {
+        state: state,
+    };
     build_display_list.traverse(flow_root);
     build_display_list.state
 }
 
-pub fn iterate_through_flow_tree_fragment_border_boxes(root: &mut Flow,
-                                                       iterator: &mut FragmentBorderBoxIterator) {
+pub fn iterate_through_flow_tree_fragment_border_boxes(root: &mut Flow, iterator: &mut FragmentBorderBoxIterator) {
     fn doit(flow: &mut Flow,
             level: i32,
             iterator: &mut FragmentBorderBoxIterator,
             stacking_context_position: &Point2D<Au>) {
         flow.iterate_through_fragment_border_boxes(iterator, level, stacking_context_position);
 
         for kid in flow::mut_base(flow).child_iter_mut() {
-            let stacking_context_position =
-                if kid.is_block_flow() && kid.as_block().fragment.establishes_stacking_context() {
-                    let margin = Point2D::new(kid.as_block().fragment.margin.inline_start, Au(0));
-                    *stacking_context_position + flow::base(kid).stacking_relative_position + margin
-                } else {
-                    *stacking_context_position
-                };
+            let stacking_context_position = if kid.is_block_flow() &&
+                                               kid.as_block().fragment.establishes_stacking_context() {
+                let mut margin = Point2D::new(kid.as_block().fragment.margin.inline_start, Au(0));
+                let fragment = kid.as_block();
+                if fragment.fragment.style.get_box().transform.0.is_some() {
+                    let relative_position = fragment
+                        .fragment
+                        .stacking_relative_border_box(&fragment.base.stacking_relative_position,
```

Comment:
Let's factor this into a helper that takes a `CoordinateSystem` argument, since it's called from multiple places of layout with the same arguments.